### PR TITLE
Add location configuration and add location horizontal cell setup

### PR DESCRIPTION
**Pull request by SSPF20**
| Demo |
| -------- |
|Simulator Screen Recording - iPhone 14 Pro - 2023-07-10 at 20 33 18|

### CODE REVIEW

**Review by juanca172 — APPROVED** [review]

**Comment by juanca172** [comment]
all work

**Review by juanca172 — APPROVED** [review]

**Review by juanca172 — APPROVED** [review]
all looks good

**Review by hgp691 — APPROVED** [review]

File: RickAndMorty/Coordinator/Coordinator.swift [hunk 1]
```diff
@@ -17,14 +17,35 @@ final class Coordinator {
     }
     
     private func setupViewControllers() {
-        let entity = RMEntity<RMCharacter, RMCharacterConfiguration>(configuration: RMCharacterConfiguration())
-        let dataProvider = RMDataProvider<RMCharacter, RMCharacterConfiguration>(entity: entity)
-        let viewModel = RMListViewModel<RMCharacter, RMCharacterConfiguration>(dataProvider: dataProvider)
-        let charactersViewController = RMListViewController(viewModel: viewModel)
-        tabBarController.viewControllers = [charactersViewController]
+        let entityCharacter = RMEntity<RMCharacter, RMCharacterConfiguration>(configuration: RMCharacterConfiguration())
```
Comment (hguzman-godaddy) [inline RickAndMorty/Coordinator/Coordinator.swift]: Fix naming

File: RickAndMorty/Coordinator/Coordinator.swift [hunk 2]
```diff
@@ -17,14 +17,35 @@ final class Coordinator {
...
+        let dataProviderLocation = RMDataProvider<RMLocation, RMLocationConfiguration>(entity: entityLocation)
+        let viewModelLocation = RMListViewModel<RMLocation, RMLocationConfiguration>(dataProvider: dataProviderLocation)
+        let locationsViewController = RMListViewController(viewModel: viewModelLocation)
+        
+        tabBarController.viewControllers = [setNavigationController(for: charactersViewController,
+                                                                    title: NSLocalizedString("Characters", comment: ""),
+                                                                    image: UIImage(systemName: "person.circle.fill")!),
+                                            setNavigationController(for: locationsViewController,
+                                                                    title: NSLocalizedString("Locations", comment: ""),
+                                                                    image: UIImage(systemName: "location.fill")!)]
```
Comment (hguzman-godaddy) [inline RickAndMorty/Coordinator/Coordinator.swift]: Do not use forced unwrap never !!! @juanca172

File: RickAndMorty/Coordinator/Coordinator.swift [hunk 3]
```diff
@@ -17,14 +17,35 @@ final class Coordinator {
...
     }
     
     func getRootViewController() -> UIViewController {
         tabBarController
     }
+    
+    private func setNavigationController(for rootViewController: UIViewController, title: String, image: UIImage) -> UIViewController {
```
Comment (hguzman-godaddy) [inline RickAndMorty/Coordinator/Coordinator.swift]: Make this image optional so you don't need to make the force unwrap

File: RickAndMorty/DataFetcher/RMDataProvider.swift [hunk 4]
```diff
@@ -86,3 +86,30 @@ struct RMCharacterConfiguration: Configuration {
         return viewModel
     }
 }
+
```
Comment (hguzman-godaddy) [inline RickAndMorty/DataFetcher/RMDataProvider.swift]: Move this to a new file

File: RickAndMorty/DataFetcher/RMDataProvider.swift [hunk 5]
```diff
@@ -86,3 +86,30 @@ struct RMCharacterConfiguration: Configuration {
...
+    
+    func getCellViewModelForEntity<T>(entity: T) -> RMItemCellViewModel? where T : Decodable {
+        guard let rmLocation = entity as? RMLocation else {
+            return nil
+        }
+        
+        let viewModel = RMLocationCellViewModel(location: rmLocation)
+        return viewModel
+    }
+
```
Comment (hguzman-godaddy) [inline RickAndMorty/DataFetcher/RMDataProvider.swift]: remove spaces

File: RickAndMorty/Views/Locations/RMLocationsViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+//
+
+import UIKit
+
+final class RMLocationsViewModel {
+    
+    let dataProvider: RMDataProvider<RMLocation, RMLocationConfiguration>
+    var locations: [[RMLocation]] = []
+    
+    init(dataProvider: RMDataProvider<RMLocation, RMLocationConfiguration> = RMDataProvider<RMLocation, RMLocationConfiguration>(entity: RMEntity<RMLocation, RMLocationConfiguration>(configuration: RMLocationConfiguration()))) {
```
Comment (hguzman-godaddy) [inline RickAndMorty/Views/Locations/RMLocationsViewModel.swift:15]: Make this method multiline

File: RickAndMorty/Views/Locations/RMLocationsViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        self.dataProvider = dataProvider
+    }
+    
+    func start() {
+        Task {
+            do {
+                let response = try await dataProvider.getPage(page: 1)
+                self.locations.append(response.results)
+            } catch {
+                print("Error loading first page \(String(describing: error))")
```
Comment (hguzman-godaddy) [inline RickAndMorty/Views/Locations/RMLocationsViewModel.swift:25]: Add TODO to be caution on the error handling for showing Error screens

File: RickAndMorty/Views/RMListView/Cells/RMCharacter/RMCharacterHorizontalCell.swift [hunk 8]
```diff
@@ -92,8 +91,8 @@ final class RMCharacterHorizontalCell: UICollectionViewCell, RMItemCell {
         NSLayoutConstraint.activate([
             container.topAnchor.constraint(equalTo: self.topAnchor, constant: 0),
             container.bottomAnchor.constraint(equalTo: self.bottomAnchor, constant: 0),
-            container.leadingAnchor.constraint(equalTo: self.leadingAnchor, constant: 0),
-            container.trailingAnchor.constraint(equalTo: self.trailingAnchor, constant: 0),
+            container.leadingAnchor.constraint(equalTo: self.leadingAnchor, constant: 20),
```
Comment (hguzman-godaddy) [inline RickAndMorty/Views/RMListView/Cells/RMCharacter/RMCharacterHorizontalCell.swift:63]: Can this be handled by the Compositional Layout ?

Comment (SSPF20) [reply RickAndMorty/Views/RMListView/Cells/RMCharacter/RMCharacterHorizontalCell.swift:63]: Yes, it can

**Comment by hguzman-godaddy** [comment]
Please add a GIF and we are ready to go

**Pull request merged**